src/jdk.incubator.concurrent/share/classes/jdk/incubator/concurrent/StructuredTaskScope.java

```diff
@@ -0,0 +1,1184 @@
+ *         Future<String> future2 = scope.fork(() -> query(right));
+ *
+ *         scope.joinUntil(deadline);
+ *
+ *         // @link substring="throwIfFailed" target="ShutdownOnFailure#throwIfFailed" :
+ *         scope.throwIfFailed(e -> new WebApplicationException(e));
+ *
+ *         // both subtasks completed successfully
+ *         String result = Stream.of(future1, future2)
+ *                 // @link substring="Future::resultNow" target="Future#resultNow" :
+ *                 .map(Future::resultNow)
+ *                 .collect(Collectors.joining(", ", "{ ", " }"));
+ *
+ *         ...
+ *     }
+ * }
+ *
+ * <h2>Extending StructuredTaskScope</h2>
+ *
+ * {@code StructuredTaskScope} can be extended, and the {@link #handleComplete(Future)
+ * handleComplete} overridden, to implement policies other than those implemented by
+ * {@code ShutdownOnSuccess} and {@code ShutdownOnFailure}. The method may be overridden
+ * to, for example, collect the results of subtasks that complete with a result and ignore
+ * subtasks that fail. It may collect exceptions when subtasks fail. It may invoke the
+ * {@link #shutdown() shutdown} method to shut down and cause {@link #join() join} to
+ * wakeup when some condition arises.
+ *
+ * <p> A subclass will typically define methods to make available results, state, or other
+ * outcome to code that executes after the {@code join} method. A subclass that collects
+ * results and ignores subtasks that fail may define a method that returns a collection of
+ * results. A subclass that implements a policy to shut down when a subtask fails may
+ * define a method to retrieve the exception of the first subtask to fail.
+ *
+ * <p> The following is an example of a {@code StructuredTaskScope} implementation that
+ * collects the results of subtasks that complete successfully. It defines the method
+ * <b>{@code results()}</b> to be used by the main task to retrieve the results.
+ *
+ * {@snippet lang=java :
+ *     class MyScope<T> extends StructuredTaskScope<T> {
+ *         private final Queue<T> results = new ConcurrentLinkedQueue<>();
+ *
+ *         MyScope() {
+ *             super(null, Thread.ofVirtual().factory());
+ *         }
+ *
+ *         @Override
+ *         // @link substring="handleComplete" target="handleComplete" :
+ *         protected void handleComplete(Future<T> future) {
+ *             if (future.state() == Future.State.SUCCESS) {
+ *                 T result = future.resultNow();
+ *                 results.add(result);
+ *             }
+ *         }
+ *
+ *         // Returns a stream of results from the subtasks that completed successfully
+ *         public Stream<T> results() {     // @highlight substring="results"
+ *             return results.stream();
+ *         }
+ *     }
+ *  }
+ *
+ * <h2><a id="TreeStructure">Tree structure</a></h2>
+ *
+ * StructuredTaskScopes form a tree where parent-child relations are established
```

Comment:
Should we mention what happens in the owner thread completes its execution and the scope's `close` method has not been called? I think that, as discussed offline, the fact that the thread will attempt to close any nested scopes when terminating is an important aspect of this API.

Reply:
The close method might be the right place for this. It has to specify that an attempt to close out of order will close the nested scopes and terminating without close is much the same. I'll see what I can do, thanks for this suggestion.